### PR TITLE
Re-adds ability to download from local mirrors

### PR DESCRIPTION
**Pull request by brejoc**
## What does this PR change?

Downloads (and GPG checks) from local mirrors are now possible again. This went away with the introduction of requests for GPG checks. No additional dependencies are needed.

## GUI diff

No difference.

## Documentation
- No documentation needed: This is a bugfix

## Test coverage
- No tests: This should already be covered.

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by brejoc** [comment]
@meaksh This PR includes the `requests_file.py`. Now in theory we could ship this as a package and there is already on in Factory. But I'm a bit hesitant to introduce an other dependency for just *one* file that is likely not going to change. What do you think?

**Comment by mcalmer** [comment]
@brejoc if a package already exists in Factory we will use it sooner or later. So it might be easier to just submit it now to SUSE:SLE-15-SP2:Update and adapt the dependencies. Otherwise it could happen that another product does it and we get file conflicts or we need to adapt later to use it.

In case we also need to submit tons of dependencies to get this working, we should re-think again.

Btw: a test is failing

**Comment by brejoc** [comment]
@mcalmer Maybe we can do this a little different. Actually we don't necessarily need to use `requests` here.

**Comment by brejoc** [comment]
Okay, the PR changed a little bit. All of the local file handling in now done directly in `repo.py` and we no longer need and adapter for requests.

File: backend/common/repo.py [hunk 1]
```diff
@@ -110,11 +110,19 @@ def get_pkg_index_raw(self) -> typing.Tuple[str, bytes]:
...
-                    self._pkg_index = cnt_fname, resp.content
-                    break
-                resp.close()
+                if packages_url.startswith("file://"):
+                    try:
+                        with open(packages_url.replace("file://", ""), "rb") as f:
+                            self._pkg_index = cnt_fname, f.read()
+                            break
+                    except:
+                        pass
```
Comment (brejoc) [inline backend/common/repo.py:119]: There is an exception handling missing.

File: backend/common/repo.py [hunk 2]
```diff
@@ -232,7 +271,57 @@ def get_release_index(self) -> typing.Dict[str, "DpkgRepo.ReleaseEntry"]:
...
+            if self.gpg_verify and not self._has_valid_gpg_signature(local_path):
+                raise GeneralRepoException("GPG verfication failed: {}".format(release_file))
+            try:
+                with open(release_file, "rb") as f:
+                    self._release = self._parse_release_index(f.read().decode("utf-8"))
+            except IOError:
+                raise GeneralRepoException("IOError while accessing file: {}".format(release_file))
+
+        # Repo format is flat
+        if self.is_flat():
```
Comment (meaksh) [inline backend/common/repo.py:305]:
```suggestion
        # Repo format is flat
        else:
```

**Review by meaksh — COMMENTED** [review]
Ok! This looks better now without adding a new handler for requests.

I've added a smalll nitpick and I would also consider refactoring the code around the execution of `gpg` CLI, to do it only from one single place with the necessary parameters instead of reduplicating it in several places.

**Comment by srbarrios** [comment]
Have we a bug to track it? It might be happening to some users with mirrors.

**Comment by srbarrios** [comment]
@paususe @nkrinner jfyi, that PR prevents the QA Maintenance automated test suite to run Ubuntu clients tests, due to the architecture we are forced to use in PRV, using a mirror for all our external content.

**Comment by paususe** [comment]
@brejoc @meaksh 
Where are we with this? Is the code complete and is it only a matter of reviewing?

**Comment by srbarrios** [comment]
@Bischoff Am I right to understand that you tested this patch in our environment and it worked?
Please confirm and notify to the @brejoc if he is not aware yet, so we can move forward on this PR.
We need this to test Ubuntu clients in our QAM environment.

**Comment by Bischoff** [comment]
> @Bischoff Am I right to understand that you tested this patch in our environment and it worked?

Correct. I tested the synchronization of Ubuntu with this pâtch in place. I did not test on a full testsuite though.

> Please confirm and notify to the @brejoc if he is not aware yet, so we can move forward on this PR.

Already done, I told him in private chat.

> We need this to test Ubuntu clients in our QAM environment.

File: backend/common/repo.py (hunk 1 above)
Comment (brejoc) [reply backend/common/repo.py]: We need to add logging in an other PR. :shrug:

File: backend/common/repo.py [hunk 3]
```diff
@@ -110,11 +110,17 @@ def get_pkg_index_raw(self) -> typing.Tuple[str, bytes]:
...
-                resp = requests.get(packages_url, proxies=self.proxies)
-                if resp.status_code == http.HTTPStatus.OK:
-                    self._pkg_index = cnt_fname, resp.content
-                    break
-                resp.close()
+                if packages_url.startswith("file://"):
+                    with open(packages_url.replace("file://", ""), "rb") as f:
+                        self._pkg_index = cnt_fname, f.read()
+                        break
+                    # TODO: Add logging in error case!
```
Comment (meaksh) [inline backend/common/repo.py:117]: Could we do the actual logging instead of having this `TODO` here?

File: backend/spacewalk-backend.changes [hunk 4]
```diff
@@ -11,6 +12,7 @@ Fri Sep 18 12:13:40 CEST 2020 - jgonzalez@suse.com
 
 - version 4.2.1-1
 - Only regenerate bootstrap repositories when linking new packages (bsc#1174636)
+>>>>>>> uyuni/master
```
Comment (meaksh) [inline backend/spacewalk-backend.changes:15]: We don't want to introduce this :smile: 
```suggestion
```

File: backend/common/repo.py (hunk 3 above)
Comment (brejoc) [reply backend/common/repo.py:117]: There is no logging at all. I'd rather do a second PR to introduce logging. Actually we need logging for several more cases here.

File: backend/spacewalk-backend.changes (hunk 4 above)
Comment (brejoc) [reply backend/spacewalk-backend.changes]: Outch! Thx!

**Review by meaksh — APPROVED** [review]
LGTM :+1:

**Pull request merged**